File: src_backup/upload/esp32.js
const fs = require('fs');
const path = require('path');
const os = require('os');
const {spawn, spawnSync} = require('child_process');
const ansi = require('ansi-string');

/**
 * Default flash addresses for ESP32 family bins.
 * Matches the ESPLoader / hardware-console.jsx triple at 0x0/0x8000/0x10000.
 * @readonly
 */
const DEFAULT_ADDRESSES = {
    bootloader: 0x0,
    partitions: 0x8000,
    firmware: 0x10000
};

const DEFAULT_BAUDRATE = 921600;
const ABORT_STATE_CHECK_INTERVAL = 100;

const ESPTOOL_PROGRESS_LINE = /Writing at 0x[0-9a-fA-F]+\s*\(\s*(\d{1,3})\s*%\s*\)/;
const ESPTOOL_ERROR_HINT = new RegExp(
    [
        'A fatal error occurred',
        'Failed to connect',
        'No serial data received',
        'Wrong boot mode',
        'Invalid head of packet'
    ].join('|'),
    'i'
);
const ESPTOOL_OK_LINE = /Hash of data verified|Hard resetting via RTS pin|Leaving\.\.\./i;

/**
 * Wrap the prebuilt esptool binary that ships with the esp32 Arduino core
 * (`tools/Arduino/packages/esp32/tools/esptool_py/<ver>/esptool[.exe]`).
 *
 * Mirrors the WebSerial + esptool-js flow used by Windify GUI's
 * hardware-console.jsx (lines 430-754), but runs on the Node side over
 * node-serialport so any client of the link server can flash an ESP32
 * without depending on Web Serial.
 */
class Esp32 {
    /**
     * @param {string} peripheralPath - serial port path (e.g. COM5 or /dev/ttyUSB0).
     * @param {object} config - flash options (chip, baudrate, addresses, flashMode...).
     * @param {string} userDataPath - link-lib user data root, used for temp bin files.
     * @param {string} toolsPath - link-lib tools root (where esptool_py lives).
     * @param {Function} sendstd - (text, normalizedProgress?) => void streaming callback.
     */
    constructor (peripheralPath, config, userDataPath, toolsPath, sendstd) {
        this._peripheralPath = peripheralPath;
        this._config = config || {};
        this._userDataPath = userDataPath;
        this._toolsPath = toolsPath;
        this._sendstd = typeof sendstd === 'function' ? sendstd : (() => {});

        this._abort = false;
        this._proc = null;
        this._tempDir = null;

        this._esptoolPath = this._resolveEsptoolBinary();
    }

    /**
     * Locate the prebuilt esptool inside `tools/Arduino/packages/esp32/tools/esptool_py/<ver>/`.
     * Falls back to `esptool` on PATH when the bundled copy is missing
     * (e.g. fresh checkout that has not run `npm run fetch` yet).
     * @returns {string} resolved esptool path - absolute when bundled, otherwise the bare binary name for PATH lookup.
     */
    _resolveEsptoolBinary () {
        const isWin = os.platform() === 'win32';
        const exeName = isWin ? 'esptool.exe' : 'esptool';
        const explicit = this._config.esptoolPath;
        if (explicit && fs.existsSync(explicit)) {
            return explicit;
        }

        const baseDir = path.join(
            this._toolsPath,
            'Arduino', 'packages', 'esp32', 'tools', 'esptool_py'
        );
        try {
            if (fs.existsSync(baseDir)) {
                const versions = fs.readdirSync(baseDir)
                    .filter(name => fs.statSync(path.join(baseDir, name)).isDirectory())
                    .sort()
                    .reverse();
                for (const ver of versions) {
                    const candidate = path.join(baseDir, ver, exeName);
                    if (fs.existsSync(candidate)) {
                        return path.resolve(candidate);
                    }
                }
            }
        } catch (err) {
            this._sendstd(
                `${ansi.yellow_dark}[esp32] esptool resolver warning: ${err.message}\n`
            );
        }
        return exeName;
    }

    abortUpload () {
        this._abort = true;
        if (this._proc && !this._proc.killed) {
            try {
                if (os.platform() === 'win32') {
                    spawnSync('taskkill', ['/pid', String(this._proc.pid), '/f', '/t']);
                } else {
                    this._proc.kill('SIGTERM');
                }
            } catch (err) {
                this._sendstd(`${ansi.red}[esp32] abort kill failed: ${err.message}\n`);
            }
        }
    }

    /**
     * Decode/copy the three bin payloads into a fresh temp directory.
     * Each entry can be { encoding:'base64', data } | { encoding:'hex', data }
     * | { path: '<absolute>' } | Buffer | string (treated as base64).
     * @param {object} bins - {bootloader, partitions, firmware}
     * @returns {object} resolved on-disk paths {bootloader, partitions, firmware}.
     */
    _writeBinsToTemp (bins) {
        if (!bins || typeof bins !== 'object') {
            throw new Error('uploadEsp32Bin requires bins payload');
        }
        const random = Math.random().toString(36);
        const rand = random.slice(2, 8);
        const id = `esp32_flash_${Date.now()}_${process.pid}_${rand}`;
        const tempRoot = path.join(this._userDataPath, 'esp32', id);
        fs.mkdirSync(tempRoot, {recursive: true});
        this._tempDir = tempRoot;

        const out = {};
        for (const key of ['bootloader', 'partitions', 'firmware']) {
            const entry = bins[key];
            if (!entry) {
                throw new Error(`uploadEsp32Bin missing ${key} bin`);
            }
            const target = path.join(tempRoot, `${key}.bin`);
            if (typeof entry === 'string') {
                fs.writeFileSync(target, Buffer.from(entry, 'base64'));
            } else if (Buffer.isBuffer(entry)) {
                fs.writeFileSync(target, entry);
            } else if (entry.path) {
                if (!fs.existsSync(entry.path)) {
                    throw new Error(`uploadEsp32Bin ${key} path not found: ${entry.path}`);
                }
                fs.copyFileSync(entry.path, target);
            } else if (entry.data) {
                const encoding = entry.encoding || 'base64';
                fs.writeFileSync(target, Buffer.from(entry.data, encoding));
            } else {
                throw new Error(`uploadEsp32Bin invalid ${key} bin payload`);
            }
            out[key] = target;
        }
        return out;
    }

    /**
     * Try to extract a 0..1 progress fraction from an esptool log fragment.
     * @param {string} text - chunk from stdout/stderr.
     * @returns {number|void} normalised progress 0..1, or void if the chunk has no usable progress marker.
     */
    _flashProgressFromText (text) {
        const m = text.match(ESPTOOL_PROGRESS_LINE);
        if (!m) {
            return;
        }
        const n = parseInt(m[1], 10);
        if (Number.isNaN(n)) {
            return;
        }
        return Math.min(1, Math.max(0, n / 100));
    }

    /**
     * Build the esptool argv (separate so tests can assert on it).
     * @param {object} files - resolved bin paths {bootloader, partitions, firmware}.
     * @returns {string[]} argv passed to esptool.
     */
    _buildArgs (files) {
        const cfg = this._config;
        const addr = Object.assign({}, DEFAULT_ADDRESSES, cfg.addresses || {});
        const args = [
            '--chip', cfg.chip || 'esp32s3',
            '--port', this._peripheralPath,
            '--baud', String(cfg.baudrate || DEFAULT_BAUDRATE),
            '--before', cfg.before || 'default_reset',
            '--after', cfg.after || 'hard_reset',
            'write_flash'
        ];
        if (cfg.eraseAll) args.push('--erase-all');
        args.push(
            '--flash_mode', cfg.flashMode || 'dio',
            '--flash_freq', cfg.flashFreq || '80m',
            '--flash_size', cfg.flashSize || 'keep'
        );
        args.push(`0x${Number(addr.bootloader).toString(16)}`, files.bootloader);
        args.push(`0x${Number(addr.partitions).toString(16)}`, files.partitions);
        args.push(`0x${Number(addr.firmware).toString(16)}`, files.firmware);
        return args;
    }

    /**
     * Pretty-print a chunk through sendstd with light ansi colouring.
     * @param {string} text - raw text from esptool.
     */
    _emit (text) {
        if (!text) return;
        let painted = text;
        if (ESPTOOL_ERROR_HINT.test(text)) {
            painted = ansi.red + text;
        } else if (ESPTOOL_OK_LINE.test(text)) {
            painted = ansi.green_dark + text;
        }
        const prog = this._flashProgressFromText(text);
        this._sendstd(painted, prog);
    }

    /**
     * Spawn esptool with the prepared bins and resolve when it finishes.
     * Resolves to 'Success' on exit code 0, 'Aborted' when the user
     * cancelled, otherwise rejects.
     * @param {object} bins - {bootloader, partitions, firmware}
     * @returns {Promise<string>} resolves to 'Success' or 'Aborted'; rejects with an Error on failure.
     */
    flashBins (bins) {
        return new Promise((resolve, reject) => {
            let files;
            try {
                files = this._writeBinsToTemp(bins);
            } catch (err) {
                return reject(err);
            }

            const args = this._buildArgs(files);
            this._sendstd(
                `${ansi.clear}[esp32] esptool ${path.basename(this._esptoolPath)} ${args.join(' ')}\n`
            );

            let proc;
            try {
                proc = spawn(this._esptoolPath, args, {windowsHide: true});
            } catch (err) {
                return reject(new Error(`Failed to spawn esptool: ${err.message}`));
            }
            this._proc = proc;

            proc.stdout.on('data', buf => this._emit(buf.toString()));
            proc.stderr.on('data', buf => this._emit(buf.toString()));

            const abortTimer = setInterval(() => {
                if (this._abort && proc && !proc.killed) {
                    try {
                        if (os.platform() === 'win32') {
                            spawnSync('taskkill', ['/pid', String(proc.pid), '/f', '/t']);
                        } else {
                            proc.kill('SIGTERM');
                        }
                    } catch (err) {
                        this._sendstd(`${ansi.red}[esp32] abort kill failed: ${err.message}\n`);
                    }
                }
            }, ABORT_STATE_CHECK_INTERVAL);

            proc.on('error', err => {
                clearInterval(abortTimer);
                this._proc = null;
                reject(new Error(`esptool spawn error: ${err.message}`));
            });

            proc.on('exit', (code, signal) => {
                clearInterval(abortTimer);
                this._proc = null;
                this._sendstd(`${ansi.clear}\r\n`);
                if (this._abort) {
                    return resolve('Aborted');
                }
                if (code === 0) {
                    return resolve('Success');
                }
                const reason = signal ?
                    `signal ${signal}` :
                    `exit code ${code === null ? 'null' : code}`;
                return reject(new Error(`esptool failed (${reason})`));
            });
        });
    }

    /**
     * Remove the temp directory created by {@link Esp32._writeBinsToTemp}.
     * Safe to call multiple times.
     */
    cleanup () {
        if (this._tempDir && fs.existsSync(this._tempDir)) {
            try {
                fs.rmSync(this._tempDir, {recursive: true, force: true});
            } catch (err) {
                this._sendstd(
                    `${ansi.yellow_dark}[esp32] cleanup warning: ${err.message}\n`
                );
            }
            this._tempDir = null;
        }
    }
}

module.exports = Esp32;
module.exports.DEFAULT_ADDRESSES = DEFAULT_ADDRESSES;
module.exports.DEFAULT_BAUDRATE = DEFAULT_BAUDRATE;

File: shell/src/upload/esp32.rs
//! Direct esptool `write_flash` of three pre-built bins. Port of `src/upload/esp32.js`.

use std::fs;
use std::io::{BufRead, BufReader};
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use base64::Engine;
use serde_json::Value;

use crate::ansi;
use crate::upload::{configure_killable, kill_tree, SendStd, UploadResult};

/// Default flash addresses (bootloader 0x0, partitions 0x8000, firmware 0x10000).
pub const DEFAULT_BOOTLOADER_ADDR: u64 = 0x0;
pub const DEFAULT_PARTITIONS_ADDR: u64 = 0x8000;
pub const DEFAULT_FIRMWARE_ADDR: u64 = 0x10000;
pub const DEFAULT_BAUDRATE: u64 = 921600;

pub struct Esp32 {
    peripheral_path: String,
    config: Value,
    user_data_path: PathBuf,
    tools_path: PathBuf,
    abort: Arc<AtomicBool>,
    temp_dir: Option<PathBuf>,
    esptool_path: PathBuf,
}

impl Esp32 {
    pub fn new(
        peripheral_path: &str,
        config: Value,
        user_data_path: &Path,
        tools_path: &Path,
    ) -> Self {
        let mut me = Self {
            peripheral_path: peripheral_path.to_string(),
            config,
            user_data_path: user_data_path.to_path_buf(),
            tools_path: tools_path.to_path_buf(),
            abort: Arc::new(AtomicBool::new(false)),
            temp_dir: None,
            esptool_path: PathBuf::new(),
        };
        me.esptool_path = me.resolve_esptool_binary();
        me
    }

    pub fn abort_flag(&self) -> Arc<AtomicBool> {
        self.abort.clone()
    }

    /// Port of `_resolveEsptoolBinary`.
    fn resolve_esptool_binary(&self) -> PathBuf {
        let exe_name = if cfg!(windows) {
            "esptool.exe"
        } else {
            "esptool"
        };
        if let Some(explicit) = self.config.get("esptoolPath").and_then(|v| v.as_str()) {
            let p = PathBuf::from(explicit);
            if p.exists() {
                return p;
            }
        }
        let base_dir = self
            .tools_path
            .join("Arduino")
            .join("packages")
            .join("esp32")
            .join("tools")
            .join("esptool_py");
        if let Ok(entries) = fs::read_dir(&base_dir) {
            let mut versions: Vec<PathBuf> = entries
                .filter_map(|e| e.ok())
                .map(|e| e.path())
                .filter(|p| p.is_dir())
                .collect();
            versions.sort();
            versions.reverse();
            for ver in versions {
                let candidate = ver.join(exe_name);
                if candidate.exists() {
                    return candidate;
                }
            }
        }
        PathBuf::from(exe_name)
    }

    /// Port of `_writeBinsToTemp`. Accepts base64 string / {encoding,data} /
    /// {path} entries.
    fn write_bins_to_temp(&mut self, bins: &Value) -> Result<[PathBuf; 3], String> {
        if !bins.is_object() {
            return Err("uploadEsp32Bin requires bins payload".to_string());
        }
        let ts = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_millis())
            .unwrap_or(0);
        let pid = std::process::id();
        let rand: String = uuid::Uuid::new_v4().simple().to_string()[..6].to_string();
        let id = format!("esp32_flash_{}_{}_{}", ts, pid, rand);
        let temp_root = self.user_data_path.join("esp32").join(&id);
        fs::create_dir_all(&temp_root).map_err(|e| e.to_string())?;
        self.temp_dir = Some(temp_root.clone());

        let keys = ["bootloader", "partitions", "firmware"];
        let mut out: Vec<PathBuf> = Vec::with_capacity(3);
        for key in keys {
            let entry = bins
                .get(key)
                .ok_or_else(|| format!("uploadEsp32Bin missing {} bin", key))?;
            let target = temp_root.join(format!("{}.bin", key));
            if let Some(s) = entry.as_str() {
                let data = base64::engine::general_purpose::STANDARD
                    .decode(s)
                    .map_err(|e| e.to_string())?;
                fs::write(&target, data).map_err(|e| e.to_string())?;
            } else if let Some(path) = entry.get("path").and_then(|v| v.as_str()) {
                if !Path::new(path).exists() {
                    return Err(format!("uploadEsp32Bin {} path not found: {}", key, path));
                }
                fs::copy(path, &target).map_err(|e| e.to_string())?;
            } else if let Some(data) = entry.get("data").and_then(|v| v.as_str()) {
                let encoding = entry
                    .get("encoding")
                    .and_then(|v| v.as_str())
                    .unwrap_or("base64");
                let bytes = if encoding == "hex" {
                    hex::decode(data).map_err(|e| e.to_string())?
                } else {
                    base64::engine::general_purpose::STANDARD
                        .decode(data)
                        .map_err(|e| e.to_string())?
                };
                fs::write(&target, bytes).map_err(|e| e.to_string())?;
            } else {
                return Err(format!("uploadEsp32Bin invalid {} bin payload", key));
            }
            out.push(target);
        }
        Ok([out[0].clone(), out[1].clone(), out[2].clone()])
    }

    fn cfg_str<'a>(&'a self, key: &str, default: &'a str) -> String {
        self.config
            .get(key)
            .and_then(|v| v.as_str())
            .unwrap_or(default)
            .to_string()
    }

    fn cfg_u64(&self, key: &str, default: u64) -> u64 {
        self.config
            .get(key)
            .and_then(|v| v.as_u64())
            .unwrap_or(default)
    }

    fn addr(&self, key: &str, default: u64) -> u64 {
        self.config
            .get("addresses")
            .and_then(|a| a.get(key))
            .and_then(|v| v.as_u64())
            .unwrap_or(default)
    }

    /// Port of `_buildArgs`.
    fn build_args(&self, files: &[PathBuf; 3]) -> Vec<String> {
        let mut args: Vec<String> = vec![
            "--chip".into(),
            self.cfg_str("chip", "esp32s3"),
            "--port".into(),
            self.peripheral_path.clone(),
            "--baud".into(),
            self.cfg_u64("baudrate", DEFAULT_BAUDRATE).to_string(),
            "--before".into(),
            self.cfg_str("before", "default_reset"),
            "--after".into(),
            self.cfg_str("after", "hard_reset"),
            "write_flash".into(),
        ];
        if self
            .config
            .get("eraseAll")
            .and_then(|v| v.as_bool())
            .unwrap_or(false)
        {
            args.push("--erase-all".into());
        }
        args.push("--flash_mode".into());
        args.push(self.cfg_str("flashMode", "dio"));
        args.push("--flash_freq".into());
        args.push(self.cfg_str("flashFreq", "80m"));
        args.push("--flash_size".into());
        args.push(self.cfg_str("flashSize", "keep"));

        let boot = self.addr("bootloader", DEFAULT_BOOTLOADER_ADDR);
        let part = self.addr("partitions", DEFAULT_PARTITIONS_ADDR);
        let fw = self.addr("firmware", DEFAULT_FIRMWARE_ADDR);
        args.push(format!("0x{:x}", boot));
        args.push(files[0].to_string_lossy().to_string());
        args.push(format!("0x{:x}", part));
        args.push(files[1].to_string_lossy().to_string());
        args.push(format!("0x{:x}", fw));
        args.push(files[2].to_string_lossy().to_string());
        args
    }

    fn flash_progress_from_text(text: &str) -> Option<f64> {
        // Match: Writing at 0x.. ( NN %)
        let lower = text;
        let idx = lower.find("Writing at 0x")?;
        let after = &lower[idx..];
        let paren = after.find('(')?;
        let close = after[paren..].find('%')?;
        let frag = &after[paren + 1..paren + close];
        let n: i64 = frag.trim().parse().ok()?;
        Some((n as f64 / 100.0).clamp(0.0, 1.0))
    }

    fn paint(text: &str) -> String {
        const ERR_HINTS: [&str; 5] = [
            "A fatal error occurred",
            "Failed to connect",
            "No serial data received",
            "Wrong boot mode",
            "Invalid head of packet",
        ];
        const OK_LINES: [&str; 3] = [
            "Hash of data verified",
            "Hard resetting via RTS pin",
            "Leaving...",
        ];
        if ERR_HINTS.iter().any(|h| text.contains(h)) {
            format!("{}{}", ansi::RED, text)
        } else if OK_LINES.iter().any(|h| text.contains(h)) {
            format!("{}{}", ansi::GREEN_DARK, text)
        } else {
            text.to_string()
        }
    }

    /// Port of `flashBins` — spawn esptool, stream, abort, resolve Success/Aborted.
    pub fn flash_bins(
        &mut self,
        bins: &Value,
        sendstd: &mut SendStd,
    ) -> Result<UploadResult, String> {
        let files = self.write_bins_to_temp(bins)?;
        let args = self.build_args(&files);
        let exe_label = self
            .esptool_path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("esptool");
        sendstd(
            &format!(
                "{}[esp32] esptool {} {}\n",
                ansi::CLEAR,
                exe_label,
                args.join(" ")
            ),
            None,
        );

        let mut cmd = std::process::Command::new(&self.esptool_path);
        cmd.args(&args)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());
        configure_killable(&mut cmd);

        let mut child = cmd
            .spawn()
            .map_err(|e| format!("Failed to spawn esptool: {}", e))?;

        // Abort watcher thread (mirrors the 100ms abort timer).
        let abort = self.abort.clone();
        let pid = child.id();
        let killed = Arc::new(AtomicBool::new(false));
        let killed2 = killed.clone();
        let watcher = std::thread::spawn(move || loop {
            if abort.load(Ordering::Relaxed) {
                #[cfg(windows)]
                {
                    let mut cmd = std::process::Command::new("taskkill");
                    cmd.args(["/pid", &pid.to_string(), "/f", "/t"]);
                    configure_killable(&mut cmd);
                    let _ = cmd.status();
                }
                #[cfg(unix)]
                unsafe {
                    libc::kill(-(pid as i32), libc::SIGTERM);
                }
                killed2.store(true, Ordering::Relaxed);
                break;
            }
            if killed2.load(Ordering::Relaxed) {
                break;
            }
            std::thread::sleep(std::time::Duration::from_millis(100));
        });

        let stdout = child.stdout.take();
        let stderr = child.stderr.take();
        if let Some(out) = stdout {
            for line in BufReader::new(out).lines().map_while(Result::ok) {
                let prog = Self::flash_progress_from_text(&line);
                sendstd(&format!("{}\n", Self::paint(&line)), prog);
            }
        }
        if let Some(err) = stderr {
            for line in BufReader::new(err).lines().map_while(Result::ok) {
                let prog = Self::flash_progress_from_text(&line);
                sendstd(&format!("{}\n", Self::paint(&line)), prog);
            }
        }

        let status = child.wait().map_err(|e| e.to_string())?;
        killed.store(true, Ordering::Relaxed);
        let _ = watcher.join();

        sendstd(&format!("{}\r\n", ansi::CLEAR), None);
        if self.abort.load(Ordering::Relaxed) {
            return Ok(UploadResult::Aborted);
        }
        match status.code() {
            Some(0) => Ok(UploadResult::Success),
            other => Err(format!("esptool failed (exit code {:?})", other)),
        }
    }

    /// Port of `cleanup`.
    pub fn cleanup(&mut self, sendstd: &mut SendStd) {
        if let Some(dir) = self.temp_dir.take() {
            if dir.exists() {
                if let Err(e) = fs::remove_dir_all(&dir) {
                    sendstd(
                        &format!("{}[esp32] cleanup warning: {}\n", ansi::YELLOW_DARK, e),
                        None,
                    );
                }
            }
        }
    }
}

/// Force the kill-tree path to be referenced so `upload::kill_tree` stays linked
/// even if all aborts go through the watcher thread.
#[allow(dead_code)]
pub fn _force_link(child: &mut std::process::Child) {
    kill_tree(child);
}
